tests/integration/pilot/revisions/uninstall_test.go

```diff
@@ -155,43 +100,69 @@ func TestUninstallPurge(t *testing.T) {
+func checkCPResourcesUninstalled(t *testing.T, cs cluster.Cluster, gvkResources []schema.GroupVersionKind, label string, purge bool) {
 	retry.UntilSuccessOrFail(t, func() error {
+		var reStrlist []string
+		var reItemList []unstructured.Unstructured
 		for _, gvk := range gvkResources {
 			resources := strings.ToLower(gvk.Kind) + "s"
 			gvr := schema.GroupVersionResource{Group: gvk.Group, Version: gvk.Version, Resource: resources}
-
-			if err := checkResourcesNotInCluster(cs, gvr, label); err != nil {
-				return err
-			}
+			reList, reStr := getRemainingResourcesCluster(cs, gvr, label)
+			reItemList = append(reItemList, reList...)
+			reStrlist = append(reStrlist, reStr...)
 		}
-		return nil
-	}, retry.Delay(time.Millisecond*100), retry.Timeout(time.Second*120))
+		return inspectRemainingResources(reItemList, reStrlist, purge)
+	}, retry.Delay(time.Millisecond*100), retry.Timeout(time.Second*150))
 }
 
-func checkResourcesNotInCluster(cs cluster.Cluster, gvr schema.GroupVersionResource, ls string) error {
+// getRemainingResourcesCluster get specific resources from the cluster
+func getRemainingResourcesCluster(cs cluster.Cluster, gvr schema.GroupVersionResource, ls string) ([]unstructured.Unstructured, []string) {
 	usList, _ := cs.Dynamic().Resource(gvr).List(context.TODO(), kubeApiMeta.ListOptions{LabelSelector: ls})
+	var remainingResources []unstructured.Unstructured
+	var staleList []string
 	if usList != nil && len(usList.Items) != 0 {
-		var stalelist []string
 		for _, item := range usList.Items {
 			// ignore IstioOperator CRD because the operator CR is not in the pruning list
 			if item.GetName() == "istiooperators.install.istio.io" {
 				continue
 			}
-			stalelist = append(stalelist, item.GroupVersionKind().String()+"/"+item.GetName())
+			remainingResources = append(remainingResources, item)
+			staleList = append(staleList, item.GroupVersionKind().String()+"/"+item.GetName())
 		}
-		if len(stalelist) == 0 {
-			return nil
+	}
+	return remainingResources, staleList
+}
+
+func inspectRemainingResources(reItemList []unstructured.Unstructured, reStrList []string, purge bool) error {
+	// for purge case we expect all resources removed
+	if purge {
+		if len(reStrList) != 0 {
+			msg := fmt.Sprintf("resources expected to be pruned but still exist in the cluster: %s",
+				strings.Join(reStrList, " "))
+			scopes.Framework.Warnf(msg)
+			return fmt.Errorf(msg)
+		}
+		return nil
+	}
+	// for other cases, we expect base component resources to be kept.
+	if len(reStrList) != 0 {
+		for _, remaining := range reItemList {
+			labels := remaining.GetLabels()
+			cn, ok := labels["operator.istio.io/component"]
+			// we don't need to check the legacy addons here because we would not install that in test anymore.
+			if ok && cn != string(name.IstioBaseComponentName) {
+				return fmt.Errorf("expect only non base component resources to still exist")
```

Comment:
do you mean "only base component resources to exist" here?

Reply:
yes, fixed